### PR TITLE
#18 documentation

### PR DESCRIPTION
**Pull request by lojzatran**
First draft of the documentation how to use the paypal-plus integration.

I also added parsing parameters as JSON for execute payment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,3 +19,61 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+paypalPlusPaymentId=${paymentId}
+```
+**NOTE - in the next few days, the logic will change and you will need to pass the parameters as JSON:**
+```
+{"paypalPlusPaymentId": "${paymentId}", "paypalPlusPayerId": "${payerId}"}
+``` 
+Example response on success:
+```json
+{
+    "statusCode": 200
```
Comment (butenkor) [inline README.md]: it should be 201.

File: README.md [hunk 2]
```diff
@@ -19,3 +19,61 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+```
+This `approval_url` should be passed to frontend to generate a list of available payment methods. Example response:
+```
+{
+    "statusCode": 201,
+    "approvalUrl": "https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=${token}"
+}
+```
+
+2. After the user pays, he/she gets redirected to the `successUrl` defined in step 1. Additionally, PayPal Plus will add 3 request parameters to this URL.
```
Comment (butenkor) [inline README.md]: 2) Execute payment after user successfully finished PayPal Plus checkout and was redirected back to the shop through `successUrl`. 

PayPal Plus will set 3 request parameters to `successUrl`:
- token
- paymentId - identifies this particular payment. **Required for execute payment.**
- PayerID - identifies the particular payer.  **Required for execute payment.**   

a) Backend extracts PayPal specific parameters: `paymentId`, `PayerID` and POSTs them to PayPal-Integration. Example:
...

**Comment by lojzatran** [comment]
Change documentation: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/cb1bacebb5bc11c1a17faac45a71ecaecdffb5a1

File: src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java [hunk 3]
```diff
@@ -28,11 +31,11 @@ public CommercetoolsExecutePaymentsController(@Nonnull StringTrimmerEditor strin
             method = RequestMethod.POST,
             value = "/{tenantName}/commercetools/execute/payments/")
     public ResponseEntity executePayments(@PathVariable String tenantName,
-                                                 @RequestParam String paypalPlusPaymentId,
-                                                 @RequestParam String paypalPlusPayerId) {
+                                          @Valid @RequestBody PaypalPlusExecuteParams params,
+                                          BindingResult result) {
```
Comment (lojzatran) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/07e2504165bbd1c392cf32bb380e10349f806bfa

File: src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java [hunk 4]
```diff
@@ -26,13 +30,15 @@ public CommercetoolsExecutePaymentsController(@Nonnull StringTrimmerEditor strin
...
         PaymentHandleResponse paymentHandleResponse = paymentHandlerProvider
                 .getPaymentHandler(tenantName)
-                .map(paymentHandler -> paymentHandler.executePayment(paypalPlusPaymentId, paypalPlusPayerId))
+                .map(paymentHandler -> paymentHandler.executePayment(params.getPaypalPlusPaymentId(), params.getPaypalPlusPayerId()))
```
Comment (butenkor) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java]: what happens if validation of parameters fails. Will the caller see nice error response?

Comment (lojzatran) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java]: If he does not provide both params, he will get back 400 and the following message. It's not the best but it provides all the info:

```json
{
    "errorMessage": "Field error in object 'paypalPlusExecuteParams' on field 'paypalPlusPayerId': rejected value [null]; codes [NotBlank.paypalPlusExecuteParams.paypalPlusPayerId,NotBlank.paypalPlusPayerId,NotBlank.java.lang.String,NotBlank]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [paypalPlusExecuteParams.paypalPlusPayerId,paypalPlusPayerId]; arguments []; default message [paypalPlusPayerId]]; default message [may not be empty]. Field error in object 'paypalPlusExecuteParams' on field 'paypalPlusPaymentId': rejected value [null]; codes [NotBlank.paypalPlusExecuteParams.paypalPlusPaymentId,NotBlank.paypalPlusPaymentId,NotBlank.java.lang.String,NotBlank]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [paypalPlusExecuteParams.paypalPlusPaymentId,paypalPlusPaymentId]; arguments []; default message [paypalPlusPaymentId]]; default message [may not be empty]"
}
```

Comment (lojzatran) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/80940626ae727b3c40dd88cfec7cf74660c136c6

File: README.md [hunk 5]
```diff
@@ -19,3 +19,50 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+    PayPal Plus will set 3 request parameters to `successUrl`:
+    - `token`
+    - `paymentId` - identifies this particular payment. **Required for execute payment.**
+    - `PayerID` - identifies the particular payer. **Required for execute payment.**
+    
+    Example of `successUrl` returned by PayPal Plus: 
+    ```
+    http://example.com/checkout/payment/success?paymentId=${paymentId}&token=${token}&PayerID=${payerId} 
+    ```
+    1. Backend extracts PayPal specific parameters: `paymentId`, `PayerID` and POSTs them to Paypal-integration. Example:
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/bdb0e44a157c291bb58b6da510db6304390bafbb

File: README.md [hunk 6]
```diff
@@ -19,3 +19,50 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+            - cancelUrl
+            - successUrl
+            - paymentMethodInfo needs to be set like this:
+            ```json
+              "paymentMethodInfo": {
+                "paymentInterface": "PAYPAL_PLUS",
+                "method": "paypal"
+              }
+            ```
+    1. Backend POSTs CTP payment ID to Paypal-integration. Example:
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/bdb0e44a157c291bb58b6da510db6304390bafbb

File: README.md [hunk 7]
```diff
@@ -19,3 +19,50 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+                "paymentInterface": "PAYPAL_PLUS",
+                "method": "paypal"
+              }
+            ```
+    1. Backend POSTs CTP payment ID to Paypal-integration. Example: 
+        ```
+        POST http://paypal-plus-integration-server.com/${tenantName}/commercetools/create/payments/${ctpPaymentId}
+        ```
+        If request was successful both response body and CTP payment object will have `approval_url` defined.
+    1. Frontend uses returned `approval_url` to render available payment methods as described in the Paypal Plus integration documentation.
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/b98999ebd75f7216bd48f1ebd0726a1586b8d126#diff-04c6e90faac2675aa89e2176d2eec7d8R51

File: README.md (hunk 1 above)
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/cb1bacebb5bc11c1a17faac45a71ecaecdffb5a1

File: README.md [hunk 8]
```diff
@@ -19,3 +19,61 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+- paymentId - identifies this particular payment. **Required for execute payment.**
+- PayerID - identifies the particular payer.  **Required for execute payment.**   
+Example of success URL returned by PayPal Plus: 
+```
+http://example.com/success/23456789?paymentId=${paymentId}token=${token}&PayerID=${payerId} 
+```
+The backend needs to call the integration server to patch and execute payment with **paymentId** and **PayerId** obtained from success URL.
+Patch will update the user's address on PayPal and execute will finalize the payment process (money is transferred). After the payment is finalized, it cannot be patched anymore.
+```
+POST http://paypal-plus-integration-server.com/${projectName}/commercetools/execute/payments/
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/cb1bacebb5bc11c1a17faac45a71ecaecdffb5a1

File: README.md [hunk 9]
```diff
@@ -19,3 +19,61 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+}
+```
+
+2. After the user pays, he/she gets redirected to the `successUrl` defined in step 1. Additionally, PayPal Plus will add 3 request parameters to this URL.
+- token
+- paymentId - identifies this particular payment. **Required for execute payment.**
+- PayerID - identifies the particular payer.  **Required for execute payment.**   
+Example of success URL returned by PayPal Plus: 
+```
+http://example.com/success/23456789?paymentId=${paymentId}token=${token}&PayerID=${payerId}
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/cb1bacebb5bc11c1a17faac45a71ecaecdffb5a1

File: README.md (hunk 2 above)
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/cb1bacebb5bc11c1a17faac45a71ecaecdffb5a1

File: README.md [hunk 10]
```diff
@@ -19,3 +19,61 @@ file. If the file exists - it overrides main context config and config from  [`/
...
 which is used for local run/debug, because the integration tests will remove all data form the CTP project**
+
+
+## How to use
+In this process, there are 3 parties involved:
+* The frontend -  the browser part of the shop. This is what the user interacts with.
+* The backend - the shop server.
+* The integration server - this is the place where this project commercetools-paypal-plus-integration is deployed and running. 
+
+1. In order to create a payment on PayPal Plus, backend must create first Payment object assigned to a cart. It must contain the following attributes:
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/cb1bacebb5bc11c1a17faac45a71ecaecdffb5a1

File: README.md [hunk 11]
```diff
@@ -19,3 +19,61 @@ file. If the file exists - it overrides main context config and config from  [`/
 
 **Note: keep this settings different from `/src/main/resources/config/application.yml`, 
 which is used for local run/debug, because the integration tests will remove all data form the CTP project**
+
+
+## How to use
+In this process, there are 3 parties involved:
+* The frontend -  the browser part of the shop. This is what the user interacts with.
+* The backend - the shop server.
+* The integration server - this is the place where this project commercetools-paypal-plus-integration is deployed and running.
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/cb1bacebb5bc11c1a17faac45a71ecaecdffb5a1

File: src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java (hunk 4 above)
Comment (butenkor) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java]: Posted error response is not user friendly and exposes unrelated internals generated by spring framework. You could skip validation by Spring and validate by yourself or if possible configure Spring to return nice error messages only like that:

`Required field 'paypalPlusPayerId' is not set.` or `Required field 'paypalPlusPaymentId' is not set paypalPlusPaymentId`

**Comment by butenkor** [comment]
Please address last comment and you are good to merge: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34#discussion_r132666000

File: src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java (hunk 4 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsExecutePaymentsController.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/issues/46

File: README.md [hunk 12]
```diff
@@ -19,3 +19,50 @@ file. If the file exists - it overrides main context config and config from  [`/
...
+            - amountPlanned
+            - cancelUrl
+            - successUrl
+            - paymentMethodInfo needs to be set like this:
+            ```json
+              "paymentMethodInfo": {
+                "paymentInterface": "PAYPAL_PLUS",
+                "method": "paypal"
+              }
+            ```
```
Comment (lojzatran) [inline README.md]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/34/commits/5bf722760ed80a298d84e677f654880489eb51c6

**Pull request merged**